scarb/src/ops/subcommands.rs

```diff
@@ -1,14 +1,87 @@
-use std::env;
+use std::collections::BTreeMap;
 use std::ffi::OsStr;
+use std::fmt::Display;
 use std::path::PathBuf;
 use std::process::Command;
+use std::{env, fs};
 
 use anyhow::{bail, Result};
 
 use crate::core::Config;
 use crate::process::{exec_replace, is_executable};
 use crate::SCARB_ENV;
 
+pub const EXTERNAL_CMD_PREFIX: &str = "scarb-";
+
+#[derive(PartialEq, Eq, PartialOrd, Ord)]
+pub enum CommandInfo {
+    BuiltIn { about: Option<String> },
+    External { path: PathBuf },
+}
+
+impl Display for CommandInfo {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            CommandInfo::BuiltIn { about: Some(about) } => write!(f, "{}", about),
+            CommandInfo::BuiltIn { about: None } => write!(f, "",),
+            CommandInfo::External { path } => write!(f, "{}", path.display()),
+        }
+    }
+}
+
+pub struct CommandsList {
+    pub commands: BTreeMap<String, CommandInfo>,
+}
+
+impl Display for CommandsList {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "Installed Commands:")?;
+        for (name, info) in self.commands.iter() {
+            writeln!(f, "{:<20}: {info}", name)?;
+        }
+        Ok(())
+    }
+}
+
+#[tracing::instrument(level = "debug", skip(config, builtins))]
+pub fn list_commands(
+    config: &Config,
+    builtins: &mut BTreeMap<String, CommandInfo>,
+) -> CommandsList {
```

Comment:
I think we might not want to provide such types/functions in scarb as a library? They seem too coupled with the way Scarb's CLI works...

I think it would be better to move this to the `bin` part